### ISSUE TITLE
Lab 6 Main.java

### ISSUE THREAD

**Issue by tjdbryan** [nus-cs2030/2223-s2#457]
Hi, anyone also got this error from Main class and managed to resolve it?
![image](https://user-images.githubusercontent.com/122286169/226157056-b6a5f07b-c219-4676-8d4e-05f69b473e27.png)

My code logic is
`class Main {
    ...
    while (sc.hasNext()) {
        for (int i = 1; i <= intValue; i++) {
            // capture student, module, assessment, grade & add them into roster
            sc.nextLine();
        }

        // capture student, module, assessment
        System.out.println(roster.getGrade(...));
        sc.nextLine();

        if (sc.next().equals("^D")) {
            sc.close();
        }
`
Thanks! 😀

**Comment by Blackknight25** [comment]
i did import java.util.NoSuchElementException then below i used the 
try{...
} catch (NoSuchElementException e) { 
break;}

**Comment by sushimakerr** [comment]
@Blackknight25 hi! i tried using the java.util.NoSuchElementException as well but codecrunch seemed to restrict the use of that. do you have an idea of how to go about without needing to use the NoSuchElementException?

**Comment by Huanqing-Mao** [comment]
Don't need to check ^D, cos ^D = control D and it's just a command, which should not be a string input? you just need to process the lines before that

**Comment by territxl** [comment]
yup, i think to avoid this error from surfacing, you can consider checking using the function Scanner.hasNext()

**Comment by leezhanpeng** [comment]
I don't think there is a need for `sc.nextLine()`. What that does is it skips the incoming input, so chaining `sc.next()` is more than sufficient. This is also why you have the error, since your code have prematurely reached the end of the input file, and is calling for next line.

**Comment by dylanlo8** [comment]
Hi! I did a check on whether sc.hasnext == true to evaluate whether the programme should cut and it works!

**Comment by tinghwei** [comment]
Your error lies in your code logic. You should not be putting the records and queries under the same loop. Records would be processed first and added to the roster before the queries can come in. As such, you should move the for loop out of the while loop so there are two different loops which should look something like this

```
for (int i = 1; i <= intValue; i++) {
    String student = sc.next();
    // repeat for module, assessment, grade & add them into roster
    sc.nextLine();
}

while (sc.hasNext()) {
    String student = sc.next();
    // repeat for module, assessment
    System.out.println(roster.getGrade(student, ...));
    sc.nextLine();
}
```

And as @Huanqing-Mao mentioned, there is no need to check for ^D.

**Comment by jeffkhong11** [comment]
If your test input eg. 1.in includes ^D, you can remove it and the error should go away on jshell, assuming that u were doing java Main < 1.in

**Comment by tjdbryan** [comment]
Thanks all!

For anyone who is facing this issue, as @jeffkhong11 mentioned, just remove ^D in your testing file. As long as your logic in Main.java is correct, the issue will not surface.

Cheers! 😀
